pageserver/benches/bench_layer_map.rs

```diff
@@ -183,64 +150,110 @@ fn bench_from_captest_env(c: &mut Criterion) {
 // Benchmark using metadata extracted from a real project that was taknig
 // too long processing layer map queries.
 fn bench_from_real_project(c: &mut Criterion) {
-    // TODO consider compressing this file
+    // Init layer map
+    let now = Instant::now();
     let layer_map = build_layer_map(PathBuf::from("benches/odd-brook-layernames.txt"));
+    println!("Finished layer map init in {:?}", now.elapsed());
+
+    // Choose uniformly distributed queries
     let queries: Vec<(Key, Lsn)> = uniform_query_pattern(&layer_map);
 
-    // Test with uniform query pattern
-    c.bench_function("real_map_uniform_queries", |b| {
+    // Choose inputs for get_difficulty_map
+    let latest_lsn = layer_map
+        .iter_historic_layers()
+        .map(|l| l.get_lsn_range().end)
+        .max()
+        .unwrap();
+    let partitioning = uniform_key_partitioning(&layer_map, latest_lsn);
+
+    // Check correctness of get_difficulty_map
+    // TODO put this in a dedicated test outside of this mod
+    {
+        println!("running correctness check");
+
+        let now = Instant::now();
+        let result_bruteforce = layer_map.get_difficulty_map_bruteforce(latest_lsn, &partitioning);
+        assert!(result_bruteforce.len() == partitioning.parts.len());
+        println!("Finished bruteforce in {:?}", now.elapsed());
+
+        let now = Instant::now();
+        let result_fast = layer_map.get_difficulty_map(latest_lsn, &partitioning);
+        assert!(result_fast.len() == partitioning.parts.len());
+        println!("Finished fast in {:?}", now.elapsed());
+
+        // Assert results are equal. Manually iterate for easier debugging.
+        //
+        // TODO The difficulty numbers are in the 50-80 range. That's pretty bad.
+        //      We should be monitoring it in prod.
```

Comment:
@knizhnik Is this surprising? I'm wondering if my test setup is wrong. I didn't expect to see 50 deltas without an image in between

Reply:
It is not surprising for me. But I consider it as not expected. With current setting we should generate new image layer after 3 delta layers. Why it doesn't happen? I do not know precisely. It has to be investigated.
I see two possible reasons for it:
1.  There is issue #2984 reported by you which cause pagestore to generate image layers for almost the whole database. In case of Kontor database is really large so generation all page images will take a lot of time. During this time a lot of new delta layers can be created.
2. Compaction and materialization are performed by one threads. So if we are busy with compaction (which can take a lot of time), we are not performing page reconstruction. And once again - it is possible that during this time a lot of delta layers will be generated.

I have not investigated much layer layout of Kontor data. I just noticed one thing - extraction of serialized directory content requires 46 layers. I found it very strange and suspicious. And it corresponds with your difficulty calculation.
That is one of the motivation of my post about storage refactoring: 
https://neondb.slack.com/archives/C033RQ5SPDH/p1673247652199539
to which I got no any reaction (looks like all reported issues are not considered as a problems and we are completely satisfied with our current layer maintenance  policy).

Reply:
Ok opened an issue. I'll reply to your slack post (I've been snoozing a slack reminder to reply to it for a while :)

Reply:
Perhaps the issue could be linked here, since it was created for this TODO?

```suggestion
        // TODO The difficulty numbers are in the 50-80 range. That's pretty bad.
        //      We should be monitoring it in prod. See: https://github.com/neondatabase/neon/issues/3316
```